### DISCUSSION TITLE
Myanmar (Burmese) language or font is missing.

### DISCUSSION THREAD

**Discussion by seth-aung** [koreader/koreader#8655] [Q&A] [answered]
Myanmar (Burmese) language or font in Koreader shows no text but squares when I use a table of contents or translate. Is there a way to make it right?
![419of83sb3881](https://user-images.githubusercontent.com/94091504/149014526-6da2d58b-ef2c-4290-914f-44f5b17dd2c1.jpg)
![80nc1ngsb3881](https://user-images.githubusercontent.com/94091504/149014544-366c78d4-4918-4254-b954-8e910e2fb1a0.jpg)

**Comment by poire-z** [comment, marked as answer]
I guess you have a Mayanmar font installed, and you see it used in EPUB/HTML/TXT books, but you miss stuff in the UI...
We don't shop https://fonts.google.com/noto/specimen/Noto+Sans+Myanmar, but you must have install it yourself.
But it would not be used by our UI elements (menu, TOC, dicts, wikipedia, translators...)

For rendering UI text, we use these fonts in that order:
https://github.com/koreader/koreader/blob/188c7d289b7048bde9cbfbe610859e07d2d60458/frontend/ui/font.lua#L105-L115
We probably should find a way to dynamically add to this list any other NotoSansXYZ from an additional list, if they are found, even if we don't ship them.

In the meantime, you can manually add in this file, in this list (if you can, on Android you might need to be rooted):
`[8] = "NotoSansMyanmar-Regular.ttf",`

**Reply by seth-aung** [reply]
Thank you for answering.

**Comment by poire-z** [comment]
Thought we could just ship `font/fallback_ui/README.txt` explaining users can drop lang specific fonts in there to be used as fallbacks for the UI, and just insert what's found in there in our `fallbacks` list above.
Which is easy on Kobo - but how this could/should work on Android and may be other platform is off putting...
frontend/fontlist.lua, frontend/ui/elements/font_settings.lua

I wouldn't want to have to kludge our CreDocument-only font menu to allow selecting some of those as UI fallbacks. A just drop/duplicate the fonts you need in this dir and forget - would feel simpler.

**Reply by pazos** [reply]
@poire-z: maybe if we put `font/fallback_ui` in dataDir? Otherwise is difficult. Not exactly android because we can export `/data/data/org.koreader.launcher/files/font/fallback_ui` and any SAF compatible FM would be able to write there without problems (on android 4.4+). But Debian/AppImage/MacOS apps are intended to be read only.

**Reply by poire-z** [reply]
This could be nice, but would need some rework to frontend/fontlist.lua, frontend/ui/elements/font_settings.lua to use this as a generic font directory (on all devices) - but then ./fonts/ and getDataDir().."/fonts" would be the same on Kobo :)
I found out I have a /sdcard/koreader/fonts/ containing my Bitter fonts - so Android has this additional trick of using this directory - I don't remember if KOReader has created it, or if I had to discover it would work and to create it myself:
https://github.com/koreader/koreader/blob/5c39c19a73d82a9687424bb202188c22e92e0219/frontend/ui/elements/font_settings.lua#L19-L20
So, we'd need to create it (like we do all the others, cache/, clipboard/...), create the fallback_ui/ subdir, and put/copy a hardcoded README.txt in it.
Still a bit off putting understanding all that and making sure there's no mess/duplicates on all platforms :)

An other option could be to just hardcode a (small) list of other non-shipped but could-be-useful-to-a-small-user-base NotoSansExoticUI.ttf, that we could pick if any has been found by FontList, whereever it was found. Hardcoding and limiting this to some handpicked NotoSans or NotoSansUI would ensure the UI stays nice and consistent, instead of picking any arbitraty MS Comic Sans Arabic the user has put in fallback_ui/ and have it screw our UI :)
This list would have to be small: we have 7 fallback fonts already, and there's a arbitrary limit of 15 in xtext.cpp. And I would have liked to insert them between NotoSansCJKsc and NotoSansArabicUI - but with too many there, we'll get many font instantiated and time wasted for nothing when we need some symbols from the late fonts.
But with such a list, I believe that on (some?) Android, all the NotoSans* are shipped, so they would be found, and used, even if the user doesn't care, with may be some symbol we would prefer to pick from our FreeSans or FreeSerif...

Still thinking :)

**Comment by seth-aung** [comment]
Hi, I've solved my problem (without root) thank to @poire-z. Let me share what I've done.

1. I opened KOreader app and go back to KOreader data dir and copied org.koreader.launcher dir and paste it somewhere in my storage. (I don't know how to get to that data/data/org.koreader.luncher dir from outside of KOreader app 😁)

2. And then open that dir via File Explorer and added NotoSansMyanmar-Regular to /file/fontend/ui/font.lua and /file/frontend/fontlist.lua using QuickEdit app like screenshots below.
![Screenshot_20220113-115550](https://user-images.githubusercontent.com/94091504/149307867-851f6d2b-4568-4bc9-943a-47e0d1c6a27c.jpg)
![Screenshot_20220113-115629](https://user-images.githubusercontent.com/94091504/149307919-a3c4e802-685e-428e-8017-c473ea392b37.jpg)

3. Then download NotoSans-Myanmar-Regular.ttf from google and added to /file/fonts/noto dir.
![Screenshot_20220113-115955](https://user-images.githubusercontent.com/94091504/149308404-34d65248-742b-4f73-afc6-c95b0955b96e.jpg)

4. Finally, I copied and replaced the whole org.koreader.luncher dir via KOreader app.

The result is like below.
![Screenshot_20220113-115257](https://user-images.githubusercontent.com/94091504/149308609-97f5b96e-99a3-4fc3-a850-fae3b2a7aae9.jpg)
![Screenshot_20220113-115337](https://user-images.githubusercontent.com/94091504/149308628-60641a0d-dac2-4967-8247-b806d7a8ede6.jpg)

Thank you for your help.

**Comment by poire-z** [comment]
Another idea:
We have https://github.com/koreader/koreader/blob/master/frontend/fontlist.lua that is require()d by frontend/ui/font, and it maintaint a list of all the fonts found in all the places allowed. We could look in that list for all fonts named "Noto Sans *", on my system, this would be:
```
01/21/22-17:26:45 DEBUG FONTNAMES ADD:  Noto Sans Myanmar
01/21/22-17:26:45 DEBUG FONTNAMES ADD:  Noto Sans Devanagari
01/21/22-17:26:45 DEBUG FONTNAMES ADD:  Noto Sans
01/21/22-17:26:45 DEBUG FONTNAMES ADD:  Noto Sans Devanagari UI
01/21/22-17:26:45 DEBUG FONTNAMES ADD:  Noto Sans CJK SC
01/21/22-17:26:45 DEBUG FONTNAMES ADD:  NotoSans Nerd Font
01/21/22-17:26:45 DEBUG FONTNAMES ADD:  Noto Sans Arabic UI
```
We could build a menu from these where the user could check those among these (we would have the ones we already use checked and not uncheckable) that he wants to see use as fallbacks (and we could limit the number of simultaneous checks to a small number like 4 or 7).
With a help/info menu item explaining where to download and drop fonts so they appear in this menu where you can enable them.

Dunno where we could put such a menu - it's not document reading related, so it may go into `Device>`, or at a first item at top or bottom of the Languages> submenu.

**Reply by seth-aung** [reply]
It will be nice. I am waiting for the next update. Thank you so much.

**Comment by poire-z** [comment]
Issue addressed with https://github.com/koreader/koreader/pull/8722 , which should avoid you to have to edit the font.lua file after each KOReader update.

I've picked some random Myanmar (I think) EPUB on the web (don't remember where and what it was), and downloaded Noto Sans Myanmar. Most characters are now shown, but there's still some random errors or blank diacritic slots. Any idea if it's the book I picked that is bad, or is Noto Sans Myanmar not that wonderful ?

![image](https://user-images.githubusercontent.com/24273478/150811601-802719ac-3010-4abc-bd7c-da2042993e92.png)

![image](https://user-images.githubusercontent.com/24273478/150811668-afd1f61e-e4d9-4451-9eb0-bd50941bae63.png)

**Reply by seth-aung** [reply]
Hi @poire-z ,
The Myanmar fonts in the screenshot above are fonts created with the Zawgyi font system used before the Myanmar Unicode system was approved in our country. The Zawgyi font system is no longer used, but some content remains unmodified.

**Comment by tjaderxyz** [comment]
Unicode provides nice HTML files of the Universal Declarations of Human
Rights in pretty much all languages:
https://unicode.org/udhr/translations.html

That is a good place to find a simple document to test a given language,
for example: https://unicode.org/udhr/d/udhr_mya.html
Message ID: ***@***.***
.com>

>

**Reply by seth-aung** [reply]
Thank you. 😊
 I have tried to put a link in reply, and it doesn't go well.

**Comment by seth-aung** [comment]
https://drive.google.com/file/d/13I61TkW-6AANyoenAmJun62jaK0TUQeu/view?fbclid=IwAR3VONQON7eFe4CwvmcIKcVLCHfJhPZE-MGcJOtWhbF0zD2mCiKdNnAAddk
This is a Myanmar book created with the Myanmar Unicode font system for you to test.

**Reply by poire-z** [reply]
My tests here are done I hope :) -I mean I think this is not a KOReader issue, but a Noto Sans Myanmar issue:
I still get some missing glyphs here and there:
![image](https://user-images.githubusercontent.com/24273478/150835982-41433014-75d9-4c0b-b39f-8480aebd72f7.png)

The Table of contents seems fine though:
![image](https://user-images.githubusercontent.com/24273478/150836119-afa3f3d0-e9e5-41dc-9183-9e10a759a23e.png)

**Reply by seth-aung** [reply]
Yes, it is not a KOreader issue or NotoSans-Myanmar issue. Missing graphics are font rendering problems caused by the ebook creator.

When you want to appear မေ, you have to type ‌ေ+မ (visual order) or မ+‌ေ(phonetic order) in a row. When deleting, it's deleted ‌ေ first and မ later. If you don't type ‌ေ+ မ or မ+‌ေ in a row, it will show like above.

e.g., you ‌accidentally type ‌ေၒ ‌, and you go back with the left arrow and type မ, it will show like ‌ေမၒ. To get perfect မေ, you need to delete both of ‌ေ‌မ and re-write that combined words in a row—that how our font works. (ebook creator didn't correct those words)

**Reply by poire-z** [reply]
OK, I understand. Thanks for explaining :)
